[translation] statically type extend(), glob(), and split()

melvinw opened the pull request:
Also, make some list functions into methods

andychu commented:
Oh BTW let's change extend() to a method, there is an AddMethods() function in core/shell.py

Same with append() and pop()

For awhile I thought we might not have methods at all, so that's why we have those functions

melvinw commented:
Pretty nice spec test bump with this one. Down to 74! http://travis-ci.oilshell.org/github-jobs/4935/cpp-spec.wwz/_tmp/spec/ysh-cpp/compare.html

andychu commented on doc/hay.md:
```diff
@@ -612,7 +612,7 @@ This can be done with an evaluator that simply enumerates all files:
       const code = parse_hay(path)
       const result = eval(hay)
-      _ append(results, result)
+      _ results->append(result)
```
Thanks for fixing these!

andychu commented on library/func_misc.py:
```diff
@@ -401,3 +421,51 @@ def Call(self, args):
+        # mycpp: rewrite list comprehension over sum type
+        l = []  # type: List[value_t]
+        for elem in self.splitter.SplitForWordEval(s, ifs=ifs):
+            l.append(value.Str(elem))
```
FYI not all  list comprehensions work in mycpp, but basic ones do

I think this can be

```
strs = self.splitter...
items = [value.Str(s) for s in strs]
return value.List(items)
```

The limitation is that it has to look lke

```
x = [...]
```

not

```
x = value.List([ ... ])  # needs a temp variable
```

melvinw replied:
Ah, yeah I had done this before. The issue was that the type checker was upset about trying to pass the temp var into the `value.List()` constructor because the tempvar was technically `List[value.Str]`. Adding a `type: List[value_t]` annotation to the tempvar fixed that, though!

andychu commented on library/func_misc.py:
```diff
@@ -401,3 +421,51 @@ def Call(self, args):
+    def __init__(self, globber):
+        # type: (glob_.Globber) -> None
+        vm._Callable.__init__(self)
+        self.globber = globber
```
Arguably we can put a Call() method directly on splitter and globber, but it doesn't matter much either way

melvinw replied:
Ah yeah, I thought about doing that, but it felt cleaner for the builtin code to be closer to the other builtins

andychu commented on library/func_misc.py:
```diff
@@ -401,3 +421,51 @@ def Call(self, args):
+        out = []  # type: List[str]
+        self.globber._Glob(s, out)
+
+        # mycpp: rewrite list comprehension over sum type
```
oh yeah I think we can do a decent job as is, as long as there's a tmp var

andychu reviewed (commented):
Looks great, thanks!

melvinw commented:
OK addressed those comments. I also tacked on a few commits to translate `shvar_get` and removed `library/func_cpython.py` while I'm in here. Have a look

andychu commented on library/func_misc.py:
```diff
@@ -37,6 +42,23 @@ def Call(self, args):
+        args.Done()
+
+        a.extend(b)
+        return value.List(a)
```
Should return `value.Null` here . 

```
>>> a=[]
>>> print(a.extend([1,2,3]))
None
```

(similar discussion with mylist.reverse(), Aidan agreed)

melvinw replied:
Ah, right! Nice catch

andychu commented on spec/ysh-builtins.test.sh:
```diff
@@ -629,3 +629,13 @@ echo $[l2[1]] # NOTE: depends on dict being ordered
+var l = List(1:3)
+echo $[len(l)]
+_ l->extend(List(3:6))
+echo $[len(l)]
```
maybe add a test for null return here, and for append(), e.g.

```
var x = l->extend(...)
if (x === null) { echo pass }
```

reminds me that we need the `assert` builtin

melvinw replied:
Added `assert()` while I was in here. What do you think?

andychu reviewed (commented):
Looks awesome, thanks!

andychu commented on spec/ysh-builtins.test.sh:
```diff
@@ -629,3 +629,31 @@ echo $[l2[1]] # NOTE: depends on dict being ordered
+setvar result = l->append(6)
+if (result === null) { echo pass }
+
+setvar result = l->pop()
```
Actually pop() is the only one that returns a value!   

```
>>> ['a', 'b'].pop()
'b'
```

I guess the original code made that mistake too ..?

melvinw replied:
Done

andychu commented:
OK thanks, I changed `assert` -> `assert_` since I think we're not quite ready to expose it

But we can use it in tests

Some threads on Zulip about the design -- proc vs. func, lazy eval, etc.

andychu commented:
Thanks, merged!
